R/estimatr_lh_robust.R

```diff
@@ -0,0 +1,184 @@
+#' @param return_vcov logical. Whether to return the variance-covariance
+#' matrix for later usage, TRUE by default.
+#' @param try_cholesky logical. Whether to try using a Cholesky
+#' decomposition to solve least squares instead of a QR decomposition,
+#' FALSE by default. Using a Cholesky decomposition may result in speed gains, but should only
+#' be used if users are sure their model is full-rank (i.e., there is no
+#' perfect multi-collinearity)
+#' @param linearHypothesis A character string or a matrix specifying combination.
+#' See \code{\link[car]{linearHypothesis}} for more details.
+#' @details
+#'
+#' This function is a wrapper for \code{\link{lm_robust}} and for
+#' \code{\link[car]{linearHypothesis}}. It first runs \code{lm_robust} and
+#' next passes \code{"lm_robust"} object as an argument to \code{linearHypothesis}.
+#'
+#' @return An object of class \code{"lh_robust"} containing the two following components:
+#'
+#' \item{lm_robust}{an object as returned by \code{lm_robust}}.
+#' \item{lh}{A data frame with most of its columns pulled from \code{linearHypothesis}' output.}
+#'
+#' The only analyis directly performed by \code{lh_robust} is a \code{t-test} for the null hypothesis of no effects of the linear combination of coefficients as specified by the user.
+#' All other output components are either extract from \code{linearHypothesis} or \code{lm_robust}.
+#'
+#' The original output returned by  \code{linearHypothesis} is added as an attribute.
+#'
+#' @examples
+#'
+#' library(fabricatr)
+#' dat <- fabricate(
+#'   N = 40,
+#'   y = rpois(N, lambda = 4),
+#'   x = rnorm(N),
+#'   z = rbinom(N, 1, prob = 0.4),
+#'   clusterID = sample(1:4, 40, replace = TRUE)
+#' )
+#'
+#' # Default variance estimator is HC2 robust standard errors
+#' lhro <- lh_robust(y ~ x + z, data = dat, linearHypothesis = "z + 2x = 0")
+#'
+#' # Also recovers other sorts of standard erorrs just as specified in \code{\link{lm_robust}}
+#' lh_robust(y ~ x + z, data = dat, linearHypothesis = "z + 2x = 0", se_type = "classical")
+#' lh_robust(y ~ x + z, data = dat, linearHypothesis = "z + 2x = 0", se_type =  "HC1")
+#'
+#' #  Can tidy() main output and subcomponents in to a data.frame
+#' lhro <- lh_robust(y ~ x + z, data = dat, linearHypothesis = "z + 2x = 0")
+#' tidy(lhro )
+#' tidy(lhro$lm_robust)
+#' tidy(lhro$lh)
+#'
+#' # Can use summary() to get more statistics on the main output and subcomponents.
+#' summary(lhro)
+#' summary(lhro$lm_robust)
+#' summary(lhro$lh)
+#'
+#' @export
+#'
+lh_robust <- function(formula,
+                      data,
+                      weights,
+                      subset,
+                      clusters,
+                      fixed_effects,
+                      se_type = NULL,
+                      ci = TRUE,
+                      alpha = .05,
+                      return_vcov = TRUE,
+                      try_cholesky = FALSE,
+                      linearHypothesis = NULL
```

Comment:
I understand why we call it `linearHypothesis`, but to fit our style it should be `linear_hypothesis`, and I would put it just after fixed_effects given that it's the main reason people would use `lh_robust`. In fact, should we even let people use `lh_robust` without `linear_hypothesis` specified?

Reply:
agreed, it should fail w/o a `linear_hypothesis`

Reply:
sounds good! just renamed it!